src/Common/src/System/Threading/Tasks/TaskToApm.cs

```diff
@@ -0,0 +1,192 @@
+            }
+            else
+            {
+                // Otherwise, the IAsyncResult should be a Task<TResult>.
+                task = asyncResult as Task<TResult>;
+            }
+
+            // Make sure we actually got a task, then complete the operation by waiting on it.
+            if (task == null)
+            {
+                throw new ArgumentNullException();
+            }
+
+            return task.GetAwaiter().GetResult();
+        }
+
+        /// <summary>Invokes the callback asynchronously when the task has completed.</summary>
+        /// <param name="antecedent">The Task to await.</param>
+        /// <param name="callback">The callback to invoke when the Task completes.</param>
+        /// <param name="asyncResult">The Task used as the IAsyncResult.</param>
+        private static void InvokeCallbackWhenTaskCompletes(Task antecedent, AsyncCallback callback, IAsyncResult asyncResult)
+        {
+            Contract.Requires(antecedent != null);
+            Contract.Requires(callback != null);
+            Contract.Requires(asyncResult != null);
+
+            // We use OnCompleted rather than ContinueWith in order to avoid running synchronously
+            // if the task has already completed by the time we get here.  This is separated out into
+            // its own method currently so that we only pay for the closure if necessary.
+            antecedent.ConfigureAwait(continueOnCapturedContext:false)
+                      .GetAwaiter()
+                      .OnCompleted(() => callback(asyncResult));
+
+            // PERFORMANCE NOTE:
+            // Assuming we're in the default ExecutionContext, the "slow path" of an incomplete
+            // task will result in four allocations: the new IAsyncResult,  the delegate+closure
+            // in this method, and the continuation object inside of OnCompleted (necessary
+            // to capture both the Action delegate and the ExecutionContext in a single object).  
+            // In the future, if performance requirements drove a need, those four 
+            // allocations could be reduced to one.  This would be achieved by having TaskWrapperAsyncResult
+            // also implement ITaskCompletionAction (and optionally IThreadPoolWorkItem).  It would need
+            // additional fields to store the AsyncCallback and an ExecutionContext.  Once configured, 
+            // it would be set into the Task as a continuation.  Its Invoke method would then be run when 
+            // the antecedent completed, and, doing all of the necessary work to flow ExecutionContext, 
+            // it would invoke the AsyncCallback.  It could also have a field on it for the antecedent, 
+            // so that the End method would have access to the completed antecedent. For related examples, 
+            // see other implementations of ITaskCompletionAction, and in particular ReadWriteTask 
+            // used in Stream.Begin/EndXx's implementation.
```

Comment:
Nit: this comment isn't relevant outside of the implementation in mscorlib, as it's referring to mscorlib internals.